ffs-shared/rule/src/commonMain/kotlin/rule/RuleEval.kt

```diff
@@ -571,19 +546,149 @@ private sealed class RuleExpr<out T> {
+        private val backing: MutableList<UByte> = mutableListOf()
+    ) : List<UByte> by backing, Comparable<Octets> {
+        init {
+            val parsedOctets = segments.ifEmpty {
+                ipString.expandIpv6().toOctets()
+            }
+            backing.addAll(parsedOctets)
+        }
+
+        override fun compareTo(other: Octets): Int {
+            if (other.size != this.size) {
+                return -1
+            }
+
+            var delta = 0u
+            for (i in other.indices) {
+                delta = this[i] - other[i]
+                if (delta != 0u) {
+                    break
+                }
             }
+
+            return delta.toInt()
+        }
+    }
+
+    internal data class IpRange(val cidr: String) : ClosedRange<Octets> {
+        private lateinit var min: Octets
+        private lateinit var max: Octets
+
+        override val endInclusive: Octets
+            get() = max
+        override val start: Octets
+            get() = min
+
+        init {
+            val ipWidth = cidr.split('/')
+            val segments = ipWidth[0].expandIpv6().toOctets()
+
+            var width = if (segments.size > 4) {
+                ipWidth.getOrNull(1)?.toShort() ?: 128
+            } else {
+                ipWidth.getOrNull(1)?.toShort() ?: 32
+            }
+
+            calculateIpRange(segments, width)
+        }
+
+        private fun calculateIpRange(octets: List<UByte>, width: Short) {
+            val min = mutableListOf<UByte>()
+            val max = mutableListOf<UByte>()
+            val fullOctets = width / 8
+            val remainder = width % 8
+            octets.forEachIndexed { i, value ->
+                val subnet: UByte = if (i < fullOctets) {
+                    (0xFF).toUByte()
+                } else if (i == fullOctets) {
+                    (0xFF shl (8 - remainder)).toUByte()
+                } else {
+                    0u
+                }
+
+                min.add(value and subnet)
+                max.add((value or subnet.inv()))
+            }
+
+            this.min = Octets(segments = min)
+            this.max = Octets(segments = max)
+        }
+    }
+
+    companion object {
+        fun String.toOctets(): List<UByte> {
+            lateinit var octets: List<UByte>
+
+            if (contains('.')) {
+                octets = split('.').map { it.toUByte() }
+                if (octets.size != 4) {
+                    throw IllegalArgumentException("invalid IPv4 format")
+                }
+            } else {
+                octets = replace(":", "")
+                    .chunked(2)
+                    .map { it.toUByte(16) }
+
+                if (octets.size != 16) {
+                    throw IllegalArgumentException("invalid IPv6 format")
+                }
+            }
+
+            return octets
+        }
+
+        fun String.expandIpv6(): String {
+            if (this.length == 39) {
+                return this
+            }
```

Comment:
I could be mistaken, but early returns can be one-liners within the official style guide.

```suggestion
            if (this.length == 39) return this
```